reV/bespoke/place_turbines.py

```diff
@@ -0,0 +1,234 @@
+                                                       yoff=-miny)
+            self.full_polygons = safe_polygons.buffer(0)
+
+            # add extra setback to cell boundary
+            minx, miny, maxx, maxy = self.full_polygons.bounds
+            minx += self.min_spacing / 2.0
+            miny += self.min_spacing / 2.0
+            maxx -= self.min_spacing / 2.0
+            maxy -= self.min_spacing / 2.0
+
+            boundary_poly = \
+                Polygon(((minx, miny), (minx, maxy), (maxx, maxy),
+                         (maxx, miny)))
+            packing_polygons = boundary_poly.intersection(self.full_polygons)
+            if isinstance(packing_polygons, MultiPolygon):
+                self.packing_polygons = packing_polygons
+            elif isinstance(packing_polygons, Polygon):
+                self.packing_polygons = MultiPolygon([packing_polygons])
+            else:
+                self.packing_polygons = MultiPolygon([])
+
+    def initialize_packing(self):
+        """run the turbine packing algorithm (maximizing plant capacity) to
+        define potential turbine locations that will be used as design
+        variables in the gentic algorithm.
+        """
+
+        packing = PackTurbines(self.min_spacing, self.packing_polygons)
+        nturbs = 1E6
+        mult = 1.0
+        while nturbs > 300:
+            packing.clear()
+            packing.min_spacing = self.min_spacing * mult
+            packing.pack_turbines_poly()
+            nturbs = len(packing.turbine_x)
+            mult *= 1.1
+        self.x_locations = packing.turbine_x
+        self.y_locations = packing.turbine_y
+
+    def optimization_objective(self, x):
+        """The optimization objective used in the bespoke optimization
+        """
+        x = [bool(y) for y in x]
+        n_turbines = np.sum(x)
+        self.wind_plant["wind_farm_xCoordinates"] = self.x_locations[x]
+        self.wind_plant["wind_farm_yCoordinates"] = self.y_locations[x]
+
+        system_capacity = n_turbines * self.turbine_capacity
+        self.wind_plant["system_capacity"] = system_capacity
+
+        self.wind_plant.assign_inputs()
+        self.wind_plant.execute()
+        aep = self.wind_plant.annual_energy()
+        cost = eval(self.cost_function, globals(), locals())
+        objective = eval(self.objective_function, globals(), locals())
+
+        return objective
+
+    def optimize(self):
+        """use a genetic algorithm to optimize wind plant layout for the user
+        defined objective function.
+        """
+        nlocs = len(self.x_locations)
+        bits = np.ones(nlocs, dtype=int)
+        bounds = np.zeros((nlocs, 2), dtype=int)
+        bounds[:, 1] = 2
+        variable_type = np.array([])
+        for _ in range(nlocs):
+            variable_type = np.append(variable_type, "int")
+        ga = GeneticAlgorithm(bits, bounds, variable_type,
+                              self.optimization_objective)
+
+        ga.max_generation = 10000
```

Comment:
This seems really sloppy. Why are these variables not set in the initialization function call `ga = GeneticAlgorithm(...)`?

Reply:
Agreed, this has been fixed. This actually threw an error when I updated '__init__' in the optimizer because it was trying to initialize a population size of zero